src/hotspot/share/opto/addnode.cpp

```diff
@@ -395,9 +395,165 @@ Node* AddNode::IdealIL(PhaseGVN* phase, bool can_reshape, BasicType bt) {
+// The method matches `n` for pattern: AddNode(a, a).
+Node* AddNode::find_simple_addition_pattern(Node* n, BasicType bt, jlong* multiplier) {
+  if (n->Opcode() == Op_Add(bt) && n->in(1) == n->in(2)) {
+    *multiplier = 2;
+    return n->in(1);
+  }
+
+  return nullptr;
+}
+
+// Try to match `a << CON`. On success, return `a` and set `1 << CON` as `multiplier`.
+// Match `n` for pattern: LShiftNode(a, CON).
+// Note that the power-of-2 multiplication optimization could potentially convert a MulNode to this pattern.
+Node* AddNode::find_simple_lshift_pattern(Node* n, BasicType bt, jlong* multiplier) {
+  // Note that power-of-2 multiplication optimization could potentially convert a MulNode to this pattern
+  if (n->Opcode() == Op_LShift(bt) && n->in(2)->is_Con()) {
+    Node* con = n->in(2);
+    if (con->is_top()) {
+      return nullptr;
+    }
+
+    *multiplier = bt == T_INT
+                    ? java_shift_left(1, con->get_int(), T_INT)
+                    : java_shift_left(1, con->get_int(), T_LONG);
+    return n->in(1);
+  }
+
+  return nullptr;
+}
+
+// Try to match `CON * a`. On success, return `a` and set `CON` as `multiplier`.
+// Match `n` for patterns:
+//     - MulNode(CON, a)
+//     - MulNode(a, CON)
+Node* AddNode::find_simple_multiplication_pattern(Node* n, BasicType bt, jlong* multiplier) {
+  // This optimization technically only produces MulNode(CON, a), but we might as match MulNode(a, CON), too.
+  if (n->Opcode() == Op_Mul(bt) && (n->in(1)->is_Con() || n->in(2)->is_Con())) {
+    Node* con = n->in(1);
+    Node* base = n->in(2);
+
+    // swap ConNode to lhs for easier matching
+    if (!con->is_Con()) {
+      swap(con, base);
+    }
+
+    if (con->is_top()) {
+      return nullptr;
+    }
+
+    *multiplier = con->get_integer_as_long(bt);
+    return base;
+  }
+
+  return nullptr;
+}
+
+// Try to match `(a << CON1) + (a << CON2)`. On success, return `a` and set `(1 << CON1) + (1 << CON2)` as `multiplier`.
+// Match `n` for patterns:
+//     - AddNode(LShiftNode(a, CON), LShiftNode(a, CON)/a)
+//     - AddNode(LShiftNode(a, CON)/a, LShiftNode(a, CON))
+// given that lhs is different from rhs.
+// Note that one of the term of the addition could simply be `a` (i.e., a << 0). Calling this function with `multiplier`
+// being null is safe.
+Node* AddNode::find_power_of_two_addition_pattern(Node* n, BasicType bt, jlong* multiplier) {
+  if (n->Opcode() == Op_Add(bt) && n->in(1) != n->in(2)) {
+    Node* lhs = n->in(1);
+    Node* rhs = n->in(2);
+
+    // swap LShiftNode to lhs for easier matching
+    if (lhs->Opcode() != Op_LShift(bt)) {
+      swap(lhs, rhs);
+    }
+
+    // AddNode(LShiftNode(a, CON), *)?
+    if (lhs->Opcode() != Op_LShift(bt) || !lhs->in(2)->is_Con()) {
+      return nullptr;
+    }
+
+    jlong lhs_multiplier = 0;
+    if (multiplier != nullptr) {
+      Node* con = lhs->in(2);
+      if (con->is_top()) {
+        return nullptr;
+      }
+
+      lhs_multiplier = bt == T_INT
```

Comment:
Why isn't it: `java_shift_left(1, con->get_int(), bt)` ?

Reply:
You're absolute right. Sorry, that was a very naive oversight.